game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/battlegroup/FiringSquadron.java

```diff
@@ -0,0 +1,241 @@
+package games.strategy.triplea.delegate.battle.steps.fire.battlegroup;
+
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Value;
+import org.triplea.java.PredicateBuilder;
+
+/** Holds predicates for finding units in a squadron and what they target */
+@Value
+@Builder(toBuilder = true)
+public class FiringSquadron {
+  /** Name that will be displayed in the Battle UI */
+  @NonNull String name;
+  /** Matches the units that will be firing */
+  @Builder.Default
+  Predicate<FiringUnitFilterData> firingUnits = FiringSquadron::defaultFiringUnitPredicate;
+  /** Matches the units that are targeted */
+  @Builder.Default
+  Predicate<TargetUnitFilterData> targetUnits = FiringSquadron::defaultTargetUnitPredicate;
+  /** This FiringSquadron can only fire if the friendly unit requirements are met */
+  @Builder.Default
+  Predicate<Collection<Unit>> friendlyUnitRequirements = FiringSquadron::defaultUnitsRequirements;
+  /** This FiringSquadron can only fire if the enemy unit requirements are met */
+  @Builder.Default
+  Predicate<Collection<Unit>> enemyUnitRequirements = FiringSquadron::defaultUnitsRequirements;
+  /** This FiringSquadron can only fire if the battle state requirements are met */
+  @Builder.Default
+  Predicate<BattleState> battleStateRequirements = FiringSquadron::defaultBattleStateRequirements;
+
+  /** ErrorProne static analysis doesn't like lambda style @Builder.Default values */
```

Comment:
I figured out a way to inline it by overriding the default lombok builder.  I did this in https://github.com/triplea-game/triplea/pull/8602/commits/bb56e98c44c6efaec443b627729f2627d646961f.

Reply:
Cool, though sorry in advance here for the churn.

I don't think it's that easy to understand this class. In part because we have a number of predicates, most of them are by default tautologies, it takes a good bit of work to know what is used, when, where and how. I think this class does not do well on the cohesion metrics where most variables are used in most methods.

What do you think of this recommendation: to split the firing squadron data from how it is created. Already there are classes I see here to create firing squadrons, they have some of the information of what is needed while this class has other parts. Having a class that is fully parameterized with generic predicates I suspect is part of the problem and makes this overly abstract (arguably a dry violation as the knowledge of how to create a firing group is split between two class, no one class is independent of the predicates from either using them or defining them).

I think it should be a goal that in this case any predicate we have should be fully encapsulated or activated via a configuration switch, but the class that uses a predicate probably should be the one that defines it. Passing in predicates I would now view as a code smell (unless we are doing a visitor pattern).